Add chapter for model comparison with large data (#174)
arviz-devs/EABM: Chapters/Model_comparison_large_data.qmd

# Model Comparison for Large Data {#sec-model-comparison-large-data}

```{python}
#| echo: false
#| warning: false
import numpy as np
import matplotlib.pyplot as plt

import arviz as az
import arviz.preview as azp
azp.style.use("arviz-variat")

plt.rcParams["figure.dpi"] = 100
SEED = 4711
```

## Introduction

Comparing Bayesian models on large datasets presents a computational challenge. Full leave-one-out cross-validation (LOO) requires computing $n$ posteriors and brute force K-fold cross-validation methods become prohibitively expensive as data size grows. 

> To overcome these challenges, we will use the approach described in @magnusson_2020 to combine fast approximations with targeted, exact computations. The key idea is to use cheap approximations, called *surrogates*, for all $n$ observations, and then correct these approximations using exact Pareto smoothed importance sampling (PSIS) LOO on just a small random subsample with a *difference estimator* that corrects for the approximation error across the full dataset.

Reviewer: ```suggestion
To overcome these challenges, we will use the approach described in @magnusson_2020. The key idea is to use cheap approximations, called *surrogates*, for all $n$ observations, and then correct these approximations using Pareto smoothed importance sampling (PSIS) LOO-CV on just a small random subsample with a *difference estimator* that corrects for the approximation error across the full dataset.
```

Predictive performance is summarized by the expected log predictive density estimated via LOO as

$$
\overline{\operatorname{elpd}}_{\text{loo}} = \frac{1}{n} \sum_{i=1}^n \pi_i,
$$

where $\pi_i = \log p(y_i\mid y_{-i})$ is the LOO predictive density for observation $i$ and $y_{-i}$ is the dataset with observation $i$ removed.

> In the example for this chapter using the `wells` data, we demonstrate the complete workflow for model comparison with subsampled PSIS-LOO. We fit a logistic regression model using both exact MCMC and Laplace approximations to show how posterior approximations can further reduce computational costs. We implement a lightweight log-likelihood helper that enables efficient evaluation on arbitrary observation subsets, demonstrate subsampled PSIS-LOO using different surrogates (PLPD and LPD), and compare multiple models with subsampled LOO.

Reviewer: ```suggestion
In the example for this chapter using the `wells` data, we demonstrate the complete workflow for model comparison with subsampled PSIS-LOO-CV. We fit a logistic regression model using both exact MCMC and Laplace approximations to show how posterior approximations can further reduce computational costs. We implement a lightweight log-likelihood helper that enables efficient evaluation on arbitrary observation subsets, demonstrate subsampled PSIS-LOO-CV using different surrogates (PLPD and LPD), and compare multiple models with subsampled LOO.
```

> In the example for this chapter using the `wells` data, we demonstrate the complete workflow for model comparison with subsampled PSIS-LOO. We fit a logistic regression model using both exact MCMC and Laplace approximations to show how posterior approximations can further reduce computational costs. We implement a lightweight log-likelihood helper that enables efficient evaluation on arbitrary observation subsets, demonstrate subsampled PSIS-LOO using different surrogates (PLPD and LPD), and compare multiple models with subsampled LOO.

Reviewer: ```suggestion
In the example for this chapter using the `wells` data, we demonstrate the complete workflow for model comparison with subsampled PSIS-LOO. We fit a logistic regression model using both MCMC sampling and Laplace approximations to show how posterior approximations can further reduce computational costs. We implement a lightweight log-likelihood helper that enables efficient evaluation on arbitrary observation subsets, demonstrate subsampled PSIS-LOO using different surrogates (PLPD and LPD), and compare multiple models with subsampled LOO.
```

## Background

The content contained in this chapter builds on foundational concepts covered in @sec-model-comparison. For detailed explanations of expected log predictive density (ELPD), leave-one-out cross-validation (LOO-CV), Pareto smoothed importance sampling (PSIS), the Watanabe-Akaike information criterion (WAIC), and model comparison principles, please refer to that chapter. Here we focus specifically on the subsampling techniques and difference estimator needed for large datasets.

### From LOO to subsampled LOO

Recall from @sec-model-comparison that PSIS-LOO uses importance sampling to approximate LOO-CV without refitting the model $n$ times. The Pareto-$k$ diagnostic flags unreliable importance sampling approximations when $k > 0.7$. For large $n$, even this efficient computation becomes expensive.

The key insight for large data is to combine three elements:

1. **Fast surrogates** $\tilde\pi_i$ computed for all $n$ observations (e.g., point predictions)

> 2. **Exact PSIS-LOO** computed only on a small random subsample $\mathcal{S}$ of size $m \ll n$

Reviewer: ```suggestion
2. **Exact PSIS-LOO-CV** computed only on a small random subsample $\mathcal{S}$ of size $m \ll n$
```

> 2. **Exact PSIS-LOO** computed only on a small random subsample $\mathcal{S}$ of size $m \ll n$

Reviewer: ```suggestion
2. **PSIS-LOO-CV** computed only on a small random subsample $\mathcal{S}$ of size $m \ll n$
```

2. Calculate $\tilde\pi_i$ for all $n$ observations using an appropriate approximation
3. Use simple random sampling to select $m$ observations (start with $m \approx 100$)
4. Calculate $\pi_j$ for the subsample using PSIS-LOO, checking Pareto-$k$ diagnostics for unreliable importance sampling approximations
5. Estimate $\operatorname{elpd}_A$, $\operatorname{elpd}_B$, and their difference
6. If subsampling SE is too large relative to the estimated difference, increase $m$ using `update_subsample`

Again, the key advantage is that the same subsample can be reused across all models, making the approach highly efficient for comparing multiple models. The difference $\operatorname{elpd}_D = \operatorname{elpd}_A - \operatorname{elpd}_B$ and its variance are estimated directly from the common subsample, properly accounting for the correlation between models' predictions. 

This reuse is what makes the difference estimator superior to the Hansen-Hurwitz approach for model comparison tasks. This is also the reason why we recommend using the same subsample for all models, rather than subsampling each model separately.

## Wells data and logistic model

> To demonstrate the PSIS-LOO subsampling method, we use data from a household survey in a region of Bangladesh where drinking water was contaminated with arsenic. Households with unsafe arsenic levels in their wells were asked whether they would switch to using a neighbor's safe well. The goal is to predict this binary switching decision using household characteristics.

Reviewer: ```suggestion
To demonstrate the PSIS-LOO-CV subsampling method, we use data from a household survey in a region of Bangladesh where drinking water was contaminated with arsenic. Households with unsafe arsenic levels in their wells were asked whether they would switch to using a neighbor's safe well. The goal is to predict this binary switching decision using household characteristics.
```
